llvm/lib/Transforms/IPO/ArgumentPromotion.cpp

```diff
@@ -611,6 +639,33 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
       // unknown users
     }
 
+    auto *CB = dyn_cast<CallBase>(V);
+    Value *PtrArg = cast<Value>(U);
+    if (CB && PtrArg && CB->getCalledFunction() == CB->getFunction()) {
```

Comment:
This should probably also check that the function type of the call and function match? I wouldn't want to reason about how the arguments map if this is not the case.

Reply:
What do you mean by function type, is it the thing returned by getFunctionType() ? I am not able to understand why it would be different ?

Reply:
The call and the function have separate types that do not match:

```
declare float @ret_float()

define i32 @call_as_i32() {
  %val = call i32 @ret_float()
  ret i32 %val
}
```

Reply:
created a new PR to address these changes, please take a look (https://github.com/llvm/llvm-project/pull/98657)